crates/nu-parser/src/hir/path.rs

```diff
@@ -39,3 +39,105 @@ impl Path {
         (self.head, self.tail)
     }
 }
+
+// WIP: find a better place for these functions.
+
+use libc::getpwnam;
+use std::error::Error;
+use std::ffi::{CStr, CString, NulError};
+use std::fmt;
+use std::path::PathBuf;
+
+type HomedirResult = Result<Option<String>, UsernameError<NulError>>;
+
+/// Matches exising user names
+///
+/// Used to resolve `~username` syntax, may fail if the username contains a null with
+/// `NulError`. Otherwise will return a `String` with the user dir (not checked to be a
+/// valid path) or `None` if there is no matching username.
+pub(crate) fn username_homedir(name: &str) -> HomedirResult {
+    let cname = CString::new(name).map_err(|e| UsernameError {
+        username: name.into(),
+        cause: e,
+    })?;
+    fn get_dir_from_username(cname: CString) -> Option<String> {
+        let pwd = unsafe { getpwnam(cname.as_ptr()) };
+        if pwd.is_null() {
+            None
+        } else {
+            let dir = unsafe {
+                let pwd = *pwd; // already checked to be not null
+                CStr::from_ptr(pwd.pw_dir).to_string_lossy().into_owned()
+            };
+            Some(dir)
+        }
+    }
+    Ok(get_dir_from_username(cname))
+}
+
+/// Represents a username lookup error.
+///
+/// This error is returned by `username_homedir()` function. The original error is
+/// provided in the `cause` field, while `username` contains the name of a user whose
+/// lookup caused the error.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct UsernameError<E> {
+    /// The name of the problematic username.
+    pub username: String,
+    /// The original error returned by the lookup function.
+    pub cause: E,
+}
+
+impl<E: fmt::Display> fmt::Display for UsernameError<E> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "error looking user '{}' up: {}",
+            self.username, self.cause
+        )
+    }
+}
+
+impl<E: Error> Error for UsernameError<E> {
+    fn description(&self) -> &str {
+        "Homedir lookup error"
+    }
+    fn cause(&self) -> Option<&dyn Error> {
+        Some(&self.cause)
+    }
+}
+
+pub(crate) fn expand_tilde_with_context<'home>(
+    path: &str,
+    homedir: &dyn Fn() -> Option<PathBuf>,
+    username_homedir: &dyn Fn(&str) -> Option<String>,
+) -> PathBuf {
+    if !path.starts_with("~") {
+        path.into()
+    } else {
+        let home = homedir();
+        match (path, home) {
+            ("~", Some(home)) => home.into(),
+            (path, Some(home)) if path.starts_with("~/") => {
+                path.replacen("~", &home.to_string_lossy(), 1).into()
```

Comment:
Open to suggestions about how to remove this `to_string_lossy`, should we support non `str` paths?

Reply:
I think in this PR I can follow the current standard, with a lossy conversion from bytes.